lib/storage/tag_filters_test.go

```diff
@@ -307,84 +331,178 @@ func TestConvertToCompositeTagFilters(t *testing.T) {
 			Value:      []byte("bar"),
 			IsNegative: false,
 			IsRegexp:   false,
 		},
+		{
+			Key:        []byte("foo"),
+			Value:      []byte("abc"),
+			IsNegative: false,
+			IsRegexp:   true,
+		},
 		{
 			Key:        []byte("__graphite__"),
 			Value:      []byte("foo.*.bar"),
 			IsNegative: false,
 			IsRegexp:   false,
 		},
+	}, [][]TagFilter{
+		{
+			{
+				Key:        []byte("\xfe\x03barfoo"),
+				Value:      []byte("abc"),
+				IsNegative: false,
+				IsRegexp:   false,
+			},
+			{
+				Key:        []byte("__graphite__"),
+				Value:      []byte("foo.*.bar"),
+				IsNegative: false,
+				IsRegexp:   false,
+			},
+		},
 	})
 
-	// A name filter with non-name filter and a graphite filter.
+	// Regexp name filter, which can be converted to non-regexp, with non-name filter.
 	f([]TagFilter{
 		{
 			Key:        nil,
 			Value:      []byte("bar"),
 			IsNegative: false,
-			IsRegexp:   false,
+			IsRegexp:   true,
 		},
 		{
 			Key:        []byte("foo"),
 			Value:      []byte("abc"),
 			IsNegative: false,
-			IsRegexp:   true,
+			IsRegexp:   false,
 		},
+	}, [][]TagFilter{
 		{
-			Key:        []byte("__graphite__"),
-			Value:      []byte("foo.*.bar"),
+			{
+				Key:        []byte("\xfe\x03barfoo"),
+				Value:      []byte("abc"),
+				IsNegative: false,
+				IsRegexp:   false,
+			},
+		},
+	})
+
+	// Multiple values regexp filter, which can be converted to non-regexp, with non-name filter.
+	f([]TagFilter{
+		{
+			Key:        nil,
+			Value:      []byte("bar|foo"),
 			IsNegative: false,
-			IsRegexp:   false,
+			IsRegexp:   true,
 		},
-	}, []TagFilter{
 		{
-			Key:        []byte("\xfe\x03barfoo"),
+			Key:        []byte("foo"),
 			Value:      []byte("abc"),
 			IsNegative: false,
 			IsRegexp:   false,
 		},
+	}, [][]TagFilter{
 		{
-			Key:        []byte("__graphite__"),
-			Value:      []byte("foo.*.bar"),
-			IsNegative: false,
-			IsRegexp:   false,
+			{
+				Key:        []byte("\xfe\x03barfoo"),
+				Value:      []byte("abc"),
+				IsNegative: false,
+				IsRegexp:   false,
+			},
+		},
+		{
+			{
+				Key:        []byte("\xfe\x03foofoo"),
+				Value:      []byte("abc"),
+				IsNegative: false,
+				IsRegexp:   false,
+			},
 		},
 	})
 
-	// Regexp name filter, which can be converted to non-regexp, with non-name filter.
+	// Two multiple values regexp filter, which can be converted to non-regexp, with non-name filter.
 	f([]TagFilter{
 		{
 			Key:        nil,
-			Value:      []byte("bar"),
+			Value:      []byte("bar|foo"),
 			IsNegative: false,
 			IsRegexp:   true,
 		},
 		{
-			Key:        []byte("foo"),
-			Value:      []byte("abc"),
+			Key:        nil,
+			Value:      []byte("abc|def"),
 			IsNegative: false,
-			IsRegexp:   false,
+			IsRegexp:   true,
 		},
-	}, []TagFilter{
 		{
-			Key:        []byte("\xfe\x03barfoo"),
-			Value:      []byte("abc"),
+			Key:        []byte("face"),
+			Value:      []byte("air"),
 			IsNegative: false,
 			IsRegexp:   false,
 		},
+	}, [][]TagFilter{
```

Comment:
This test case looks incorrect. The `{__name__=~"foo|bar",__name__=~"abc|def", ...}` must match empty time series, since there are no series with names simultaneously matching `foo or bar` and `abc or def`.

Reply:
So if we find a duplicate `__name__` tagfilter, is it okay if we return empty tagfilers ? Similarly, there should be no metric name equal to both `foo` and `bar`

This test case also needs to be modified
![image](https://user-images.githubusercontent.com/3659110/132942037-93c98b6a-7cd9-412d-b6df-b4c28aec5807.png)

Reply:
I realized that if the duplicate `__name__` filter also contains a duplicate value, it should also need to work, and I modified it as you wrote it before.
It may be possible to get queries like `{__name__=~"foo|abc",__name__=~"foo|def", ...}` to work properly.